cirq-core/cirq/ops/projector_test.py

```diff
@@ -0,0 +1,161 @@
+import numpy as np
+import pytest
+
+import cirq
+
+
+def test_projector_qid():
+    q0 = cirq.NamedQubit('q0')
+
+    zero_projector = cirq.ProjectorString({q0: 0})
+    one_projector = cirq.ProjectorString({q0: 1})
+
+    np.testing.assert_allclose(zero_projector.matrix().toarray(), [[1.0, 0.0], [0.0, 0.0]])
+    np.testing.assert_allclose(one_projector.matrix().toarray(), [[0.0, 0.0], [0.0, 1.0]])
+
+
+def test_projector_repr():
+    q0 = cirq.NamedQubit('q0')
+
+    assert (
+        repr(cirq.ProjectorString({q0: 0}))
+        == "cirq.ProjectorString(projector_dict={cirq.NamedQubit('q0'): 0},coefficient=(1+0j))"
+    )
+
+
+def test_projector_from_np_array():
+    q0 = cirq.NamedQubit('q0')
+
+    zero_projector = cirq.ProjectorString({q0: 0})
+    np.testing.assert_allclose(zero_projector.matrix().toarray(), [[1.0, 0.0], [0.0, 0.0]])
+
+
+def test_projector_matrix_missing_qid():
+    q0, q1 = cirq.LineQubit.range(2)
+    proj = cirq.ProjectorString({q0: 0})
+
+    np.testing.assert_allclose(proj.matrix().toarray(), np.diag([1.0, 0.0]))
+    np.testing.assert_allclose(proj.matrix([q0]).toarray(), np.diag([1.0, 0.0]))
+    np.testing.assert_allclose(proj.matrix([q1]).toarray(), np.diag([1.0, 1.0]))
+
+    np.testing.assert_allclose(proj.matrix([q0, q1]).toarray(), np.diag([1.0, 1.0, 0.0, 0.0]))
+    np.testing.assert_allclose(proj.matrix([q1, q0]).toarray(), np.diag([1.0, 0.0, 1.0, 0.0]))
+
+
+def test_equality():
+    q0 = cirq.NamedQubit('q0')
+
+    obj1 = cirq.ProjectorString({q0: 0})
+    obj2 = cirq.ProjectorString({q0: 1})
+
+    assert obj1 == obj1
+    assert obj1 != obj2
+    assert hash(obj1) == hash(obj1)
+    assert hash(obj1) != hash(obj2)
+
+
+def test_get_values():
+    q0 = cirq.NamedQubit('q0')
+    d = cirq.ProjectorString({q0: 0}, 1.23 + 4.56j)
+
+    assert len(d.projector_dict) == 1
+    assert d.projector_dict[q0] == 0
+    assert d.coefficient == 1.23 + 4.56j
+
+
+def test_expectation_from_state_vector_basis_states_empty():
+    q0 = cirq.NamedQubit('q0')
+    d = cirq.ProjectorString({})
+
+    np.testing.assert_allclose(d.expectation_from_state_vector(np.array([1.0, 0.0]), {q0: 0}), 1.0)
+
+
+def test_expectation_from_state_vector_basis_states_single_qubits():
+    q0 = cirq.NamedQubit('q0')
+    d = cirq.ProjectorString({q0: 0})
+
+    np.testing.assert_allclose(d.expectation_from_state_vector(np.array([1.0, 0.0]), {q0: 0}), 1.0)
+    np.testing.assert_allclose(d.expectation_from_state_vector(np.array([0.0, 1.0]), {q0: 0}), 0.0)
+
+
+def test_expectation_from_state_vector_basis_states_three_qubits():
```

Comment:
Nit: Can a test like similar to this be added for density matrix expectation calculation ?

Reply:
Added a test in `test_projector_matrix_missing_qid` and in `test_projector_matrix`. I also decided to add yet another test: `test_consistency_state_vector_and_density_matrix` because your comment made me find a bug.